Compression support

jgrewe opened the pull request:
addresses #510 

this is more a quick and dirty implementation of compression support. It uses a hdf5 built-in compression algorithm with a fixed compression level. At the moment, one can decide for each dataArray to compress it or not, default it true.

This pr is meant to get us thinking about how it should actually work. For the moment, I would prefer  such a simple solution which could be elaborated in future versions, or upon user request...

achilleas-k commented:
Is there a reason we want compression defined for each DataSet separately? I imagine it would be nicer to have the option to enable during file creation and then each DataArray (or polynomial coefficient) creation method could check the global setting.

jgrewe commented:
@achilleas-k, well I first thought so as well but this would mean that we have to ask, e.g. ```File``` which mode was selected upon each creation of a ```DataSet```.... We can do it, though. I would keep the option to override the default and decide when not to compress, or vice versa. For example the imaging guys always keep the raw images uncompressed (well, image compression is something different, but there may be cases in which this is desired), or there are scenarios in which, e.g. for performance reasons, you want to have one ```DataArray``` uncompressed while you compress the others.

achilleas-k commented:
Yeah, I see. I can see how this might be useful.

Would it be too complicated if the user could specify the default value globally? In other words, when opening or creating a file, you could specify the true or false for compression, then the DataArray creation function accepts an argument that overrides that default.

Might be a bit messy though.

jgrewe commented:
I was wondering if such a default compression choice should be saved within the file? That is, when a file is re-opened do I automatically have the same settings? Such a feature would require a model change, however.

achilleas-k commented:
I guess it could be a top-level attribute (alongside file format and version), the absence of which signals "off". That way it would be backwards compatible in the file format.

jgrewe commented:
@achilleas-k tried to define a default during file opening. Unfortunately it won't work without an api change and this is something I would like to avoid.
We can set the default compression choice at the level of the File and pass it to the Block(s) which then can apply it at any place (in principle). The problem starts at the ````createDataArray``` function. If we want to avoid an api break, we have to add the compression flag with a default value. This is then leads to the situation that we always have a compression argument (with the default value) and a  maybe conflicting compression flag set at file opening. Maybe I am to blind, but I do not see a simple way to tell/decide which one  is the current user's wish. We could spend a further argument on this, but this becomes really ugly.

I would suggest to use the approach used here and default compression to a certain value that can be overridden at any place a dataset is created. For future versions that may break the the api anyway we may elaborate on this. Actually, I would really like to have the compression for our recordings, saves about 60% of disc space :)

achilleas-k commented:
I see. Yeah, I guess we can think about it again in the future. The global option would be in addition to the current change anyway, so we can go ahead with this one and work on it later (or not).

achilleas-k commented on backend/fs/DataArrayFS.cpp:
```diff
@@ -240,7 +240,7 @@ DataArrayFS::~DataArrayFS() {
 
 
-void DataArrayFS::createData(DataType dtype, const NDSize &size) {
+    void DataArrayFS::createData(DataType dtype, const NDSize &size, bool compression) {
```
Indent typo?

achilleas-k commented on backend/hdf5/DataArrayHDF5.cpp:
```diff
@@ -131,13 +131,13 @@ vector<double> DataArrayHDF5::polynomCoefficients() const {
         ds.setExtent({coefficients.size()});
     } else {
-        ds = group().createData("polynom_coefficients", H5T_NATIVE_DOUBLE, {coefficients.size()});
+         ds = group().createData("polynom_coefficients", H5T_NATIVE_DOUBLE, {coefficients.size()}, {}, {}, true, true, compression);
```
Another one?

jgrewe replied:
eh, fixed those ones, will rebase the branch once the other #696 is merged...

gicmo commented on backend/hdf5/h5x/H5Group.cpp:
```diff
@@ -217,8 +218,12 @@ DataSet H5Group::createData(const std::string &name,
+    DataSet ds;
+    if (compression) {
+         HErr status = H5Pset_deflate (dcpl.h5id(), 6);
+         status.check("Could not set compression!");
```
Which exception will this result in? Maybe we want to differentiate this use case client-side, so we a program could try to use compression, and if fails retry without it?

gicmo commented on backend/hdf5/DataArrayHDF5.cpp:
```diff
@@ -131,13 +131,13 @@ vector<double> DataArrayHDF5::polynomCoefficients() const {
         ds.setExtent({coefficients.size()});
     } else {
-        ds = group().createData("polynom_coefficients", H5T_NATIVE_DOUBLE, {coefficients.size()});
+        ds = group().createData("polynom_coefficients", H5T_NATIVE_DOUBLE, {coefficients.size()}, {}, {}, true, true, compression);
```
We could (should?) re-order the arguments so we can make use of the default args most of the time. I am not sure about the rest of the codebase but I see a few new `{}, {}, true, true, compression` places. We could avoid that by moving the compression arg up behind size, but maybe then other places need more default args, not sure. Do you have an idea @jgrewe ? The `createData` is internal API so I think we can do that without too much hassle, if we wanted to.

gicmo commented on backend/fs/BlockFS.cpp:
```diff
@@ -250,7 +250,7 @@ bool BlockFS::deleteSource(const std::string &name_or_id) {
 
 std::shared_ptr<base::IDataArray> BlockFS::createDataArray(const std::string &name, const std::string &type,
-                                                           nix::DataType data_type, const NDSize &shape) {
+                                                           nix::DataType data_type, const NDSize &shape, bool compression) {
```
Although I indeed very much like the simplicity of the approach, maybe we can be a tiny bit more extensible and future-proof by having an enum instead of the bool, ie:
```c++
enum class Compression {
   None,
   DeflateNormal,

   Auto = DeflateNormal,
   //and later:
   DeflateHigh,
   DeflateLow,

    FileDefault, // use the default store within the file
}
```

jgrewe commented:
@achilleas-k @gicmo, reworked the compression using @gicmo's enum approach. We can now pass it as either as a default while opening the file, or at each DataArray creation
Switched the default to NO compression

gicmo commented on backend/hdf5/BlockHDF5.cpp:
```diff
@@ -290,15 +292,16 @@ shared_ptr<ITag> BlockHDF5::createTag(const std::string &name, const std::string
 
-    // now create the actual H5::DataSet
-    da->createData(data_type, shape);
+    // now create the actual H5::Data Set
```
Nitpick: `H5::DataSet` was correct ;)

jgrewe replied:
args, struggling with the emacs settings, in particular the indentation... Thought I found them all before pushing, apparently not.

gicmo replied:
I think I fucked the emacs stuff up for us, fixating...

gicmo commented on backend/hdf5/DataArrayHDF5.cpp:
```diff
@@ -131,13 +131,13 @@ vector<double> DataArrayHDF5::polynomCoefficients() const {
         ds.setExtent({coefficients.size()});
     } else {
-        ds = group().createData("polynom_coefficients", H5T_NATIVE_DOUBLE, {coefficients.size()});
+         ds = group().createData("polynom_coefficients", H5T_NATIVE_DOUBLE, {coefficients.size()}, compression);
```
nitpick: did a space smuggle itself in?

gicmo replied:
Also more importantly: you really think it is necessary to compress the polynom dataset? Shouldn't those be only a few entries at all?

jgrewe replied:
actually not, that is because the header defines the default to be Compression::None

gicmo commented on backend/hdf5/h5x/H5Group.cpp:
```diff
@@ -217,8 +218,12 @@ DataSet H5Group::createData(const std::string &name,
+    if (compression == Compression::DeflateNormal) {
+         HErr status = H5Pset_deflate (dcpl.h5id(), 6);
+         status.check("Could not set compression!");
+    }
```
Hmm, maybe it would be good to already add the case where an unsupported compression is specified (i.e. throw an exception)?

jgrewe replied:
I think that is exactly what ``check``does, As I see it, maybe I am wrong, ``HErr::check`` throws an [H5Error](https://github.com/G-Node/nix/blob/f352378c222c16d24e01f5c1f79398fb52fd1151/backend/hdf5/h5x/H5Object.hpp#L199)

jgrewe replied:
or would you prefer an extra exception for just this case?

gicmo replied:
What I meant is that, if we add a Compression::XXX in the future, we would just ignore it in that function here, right?

gicmo commented on backend/hdf5/h5x/H5Group.hpp:
```diff
@@ -47,14 +48,15 @@ class NIXAPI H5Group : public LocID {
 
     template<typename T>
-    void setData(const std::string &name, const T &value);
+         void setData(const std::string &name, const T &value, const Compression &compression = Compression::Auto);
```
nitpick: new extra whitespaces?

gicmo commented on include/nix/Compression.hpp:
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2013, German Neuroinformatics Node (G-Node)
```
nitpick: 2013 was quite some time ago ...

jgrewe commented:
a little rebase also from my side :)
